File: cpp/DBBD/DBBD/TcpSessionBase.cpp
#include <boost/bind.hpp>
#include "TcpSessionBase.h"
#include "Serialize.h"

DBBD::TcpSessionBase::TcpSessionBase(IoContextSP context, SocketSP socket, unsigned int bufferSize)
{
	this->context = context;
	this->socket = socket;
	this->readBuffer = new Buffer(bufferSize);
	this->writeBuffer = new Buffer(bufferSize);
}

DBBD::TcpSessionBase::~TcpSessionBase()
{
	stop();
}

void DBBD::TcpSessionBase::start()
{
	startInternal();
	read();
}

void DBBD::TcpSessionBase::stop()
{
	if (isDisposed) { return; }
	isDisposed = true;

	if (socket && socket->is_open()) {
		socket->close();
	}

	if (readBuffer != nullptr) {
		delete readBuffer;
		readBuffer = nullptr;
	}

	if (writeBuffer != nullptr) {
		delete writeBuffer;
		writeBuffer = nullptr;
	}

	if (bindStopInternal) {
		bindStopInternal(sessionId);
	}
}

void DBBD::TcpSessionBase::send(DBBD::Cell* data)
{
	Serialize::write(*writeBuffer, data);
	write();
}

void DBBD::TcpSessionBase::bindingStopInternal(std::function<void(const size_t&)> method)
{
	bindStopInternal = method;
}

void DBBD::TcpSessionBase::read()
{
	readBuffer->adjust();
	if (socket && socket->is_open()) {
		socket->async_read_some(boost::asio::buffer(readBuffer->getBuffer(), 8192),
			boost::bind(&TcpSessionBase::handleRead, this,
				boost::asio::placeholders::error, boost::asio::placeholders::bytes_transferred));
	}
}

void DBBD::TcpSessionBase::handleRead(const boost::system::error_code& error, unsigned int bytesTransferred)
{
	if (error) {
		stop();
		return;
	}

	readBuffer->increaseLastPos(bytesTransferred);
	while (true) {
		if (readBuffer->getBufferLastPos() < HeaderSize) {
			break;
		}

		auto headerBlock = readBuffer->viewByteBlock(HeaderSize);
		Header header(headerBlock);

		if (readBuffer->getBufferLastPos() < header.length) {
			break;
		}

		readInternal(header);
	}

	read();
}

void DBBD::TcpSessionBase::write()
{
	writeInternal();

	if (socket && socket->is_open()) {
		boost::asio::async_write(*socket,
			boost::asio::buffer(writeBuffer->getBuffer(), writeBuffer->getBufferLastPos()),
			boost::bind(&TcpSessionBase::handleWrite, this,
				boost::asio::placeholders::error, boost::asio::placeholders::bytes_transferred));
	}
}

void DBBD::TcpSessionBase::handleWrite(const boost::system::error_code& error, unsigned int bytesTransferred)
{
	if (error) {
		stop();
		return;
	}
	
	writeBuffer->setBufferOffset(writeBuffer->getBufferOffset() + bytesTransferred);
	writeBuffer->adjust();
}

File: csharp/DBBD/DBBD/Network/TcpSessionBase.cs
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Text;

namespace DBBD
{
    public abstract class TcpSessionBase : SessionBase
    {
        public TcpSessionBase(Socket socket, uint bufferSize) 
        {
            this.socket = socket;
            readBuffer = new Buffer(bufferSize);
            writeBuffer = new Buffer(bufferSize);

            readEventArgs = new SocketAsyncEventArgs();
            sendEventArgs = new SocketAsyncEventArgs();

            readEventArgs.SetBuffer(readBuffer.RawBuffer, 0, (int)bufferSize);

            readEventArgs.Completed += ReadEventArgs_Completed;
            sendEventArgs.Completed += SendEventArgs_Completed;
        }

        public override void Start()
        {
            StartInternal();
            Read();
        }

        public override void Stop()
        {
            if (isDisposed) { return; }
            isDisposed = true;

            if(socket != null && socket.Connected)
            {
                socket.Close();
            }
        }

        public override void Send(ICell data)
        {
            Serialize.Write(writeBuffer, data);
            Write();
        }

        protected override void Read()
        {
            readBuffer.Adjust();
            if(socket != null && socket.Connected)
            {
                bool pending = socket.ReceiveAsync(readEventArgs);
                if (!pending)
                {
                    OnReceive(readEventArgs);
                }
            }
        }

        protected override void Write()
        {
            WriteInternal();
            sendEventArgs.SetBuffer(writeBuffer.RawBuffer, 0, (int)writeBuffer.BufferLastPos);
            bool pending = socket.SendAsync(sendEventArgs);
            if (!pending)
            {
                OnSend(sendEventArgs);
            }
        }

        private void SendEventArgs_Completed(object sender, SocketAsyncEventArgs e)
        {
            OnSend(e);
        }

        private void ReadEventArgs_Completed(object sender, SocketAsyncEventArgs e)
        {
            OnReceive(e);
        }

        private void OnReceive(SocketAsyncEventArgs e)
        {
            if(e.SocketError == SocketError.Success)
            {
                readBuffer.BufferLastPos += (uint)e.BytesTransferred;
                while(true)
                {
                    if(readBuffer.BufferLastPos < Header.HeaderSize)
                    {
                        break;
                    }

                    byte[] headerBlock = readBuffer.ViewByteBlock(Header.HeaderSize);
                    Header header = new Header(headerBlock);

                    if(readBuffer.BufferLastPos < header.Length)
                    {
                        break;
                    }

                    ReadInternal(header);
                    break;
                }
                Read();
            }
            else
            {
                Stop();
            }
        }

        private void OnSend(SocketAsyncEventArgs e)
        {
            if(e.SocketError == SocketError.Success)
            {
                writeBuffer.BufferOffset += (uint)e.BytesTransferred;
                writeBuffer.Adjust();
            }
            else
            {
                Stop();
            }
        }

        private SocketAsyncEventArgs readEventArgs;
        private SocketAsyncEventArgs sendEventArgs;
    }
}